associated-token-account/program/src/processor.rs

```diff
@@ -56,6 +72,26 @@ pub fn process_create_associated_token_account(
     let spl_token_program_info = next_account_info(account_info_iter)?;
     let spl_token_program_id = spl_token_program_info.key;
 
+    if create_mode == CreateMode::Idempotent
+        && associated_token_account_info.owner == spl_token_program_id
+    {
+        let ata_data = associated_token_account_info.data.borrow();
+        if let Ok(associated_token_account) = StateWithExtensions::<Account>::unpack(&ata_data) {
+            if associated_token_account.base.owner != *wallet_account_info.key {
+                let error = AssociatedTokenAccountError::InvalidOwner;
+                msg!("{}", error);
+                return Err(error.into());
+            }
+            if associated_token_account.base.mint != *spl_token_mint_info.key {
+                return Err(ProgramError::InvalidAccountData);
+            }
```

Comment:
is there any reason we don't check `associated_token_address == *associated_token_account_info.key`? maybe to save compute? strictly speaking this is breaking the idempotent? (i.e. you can get false-positive (Ok from CreateIdempotent) for arbitrary existing non-ata accounts)

Reply:
That's a great point, I'll move that check up